source/common/grpc/google_grpc_creds_impl.cc

```diff
@@ -0,0 +1,165 @@
+      if (credential.has_channel_credentials() &&
+          credential.channel_credentials().has_ssl_credentials()) {
+        const grpc::SslCredentialsOptions ssl_creds = {
+            .pem_root_certs = Config::DataSource::read(
+                credential.channel_credentials().ssl_credentials().root_certs(), true),
+            .pem_private_key = Config::DataSource::read(
+                credential.channel_credentials().ssl_credentials().private_key(), true),
+            .pem_cert_chain = Config::DataSource::read(
+                credential.channel_credentials().ssl_credentials().cert_chain(), true),
+        };
+        return grpc::SslCredentials(ssl_creds);
+      }
+    }
+    return creds;
+  }
+
+  std::string name() const override { return "envoy.grpc_credentials.default"; }
+};
+
+/**
+ * Access token implementation of Google Grpc Credentials Factory
+ * This implementation uses ssl creds for the grpc channel if available, similar to the default
+ * implementation. Additionally, it uses MetadataCredentialsFromPlugin to add a static secret to a
+ * header for call credentials. This implementation does the same thing as AccessTokenCredentials,
+ * but it's implemented as a plugin to show how a custom implementation would be created.
+ *
+ * This implementation uses the access_token field in the config to get the secret to add to the
+ * header.
+ *
+ * This can be used as an example for how to implement a more complicated custom call credentials
+ * implementation. Any blocking calls should be performed in the
+ * MetadataCredentialsFromPlugin::GetMetadata to ensure that the main thread is not blocked while
+ * initializing the channel.
+ */
+class AccessTokenGrpcCredentialsFactory : public GoogleGrpcCredentialsFactory {
+
+public:
+  std::shared_ptr<grpc::ChannelCredentials>
+  getChannelCredentials(const envoy::api::v2::core::GrpcService& grpc_service_config) override {
+    const auto& google_grpc = grpc_service_config.google_grpc();
+    std::shared_ptr<grpc::ChannelCredentials> creds =
+        grpc::SslCredentials(grpc::SslCredentialsOptions());
+    std::shared_ptr<grpc::CallCredentials> call_creds = nullptr;
+    for (const auto& credential : google_grpc.credentials()) {
+      switch (credential.credential_specifier_case()) {
+      case envoy::api::v2::core::GrpcService::GoogleGrpc::Credentials::kChannelCredentials: {
+        if (credential.channel_credentials().has_ssl_credentials()) {
+          const grpc::SslCredentialsOptions ssl_creds = {
+              .pem_root_certs = Config::DataSource::read(
+                  credential.channel_credentials().ssl_credentials().root_certs(), true),
+              .pem_private_key = Config::DataSource::read(
+                  credential.channel_credentials().ssl_credentials().private_key(), true),
+              .pem_cert_chain = Config::DataSource::read(
+                  credential.channel_credentials().ssl_credentials().cert_chain(), true),
```

Comment:
This SSL cred setup logic seems like it could be refactored as it's a dupe of above.